configure.ac

```diff
@@ -416,6 +416,8 @@ case "$host_os" in
         ;;
     *solaris*)
         AC_DEFINE_UNQUOTED(ON_SOLARIS, 1, Compiling for Solaris platform)
+        # get getpwnam_r declaration right
+        CPPFLAGS="$CPPFLAGS -D_POSIX_PTHREAD_SEMANTICS"
```

Comment:
Looking at this again, there's no reason we use getpwnam_r() instead of getpwnam(). Most likely it was added when pacemaker was trying to be thread-safe, which it never was and now explicitly rejects. I'd rather swap out the call than work around it here.

Reply:
Good catch.

Reply:
Hmm, actually, it's reasonable to use `getpwnam_r` since `getpwnam`
is not async-signal-safe, and either:

- we can inflict resulting "bad user selection" on our own if we call,
  e.g. `crm_user_lookup` from a signal handler (bad idea on its own)
  while the main code flow is exactly within `getpwnam`, specifically,
  once the assignments to the respective static variable are done)

- ditto because of hidden signal handlers of the underlying libraries
  (outside of pacemaker) or such a library can be multithreaded while
  (incorrectly) using mere `getpwnam`

So would be for actually retaining `getpwnam_r` as was on that very
place, while noting the necessesity to review async-signal safety in
the long term TODO (looks like another very neglected topic).

Reply:
I'm not sure I follow what you're saying here. By that logic, non-async-signal-safe functions should never be used. I think all we need to verify is that our own signal handlers don't call such functions.

For any of our code that uses libcrmcommon's mainloop signal handling functions, registering a "signal handler" actually registers a mainloop trigger handler, and all the real signal handler (mainloop_signal_handler()) does is set the trigger, so it's async-signal-safe. (BTW I just realized I've been setting variables as volatile unnecessarily.)

That leaves crm_signal(), which appears to be used only for crm_trigger_blackbox() when receiving a fatal signal. That has theoretical failure cases but none we need to be concerned about here. (I'm ignoring cibmon/crm_mon as being relatively unimportant in this respect.)

Reply:
1. good catch that the actual async signal handling is quite locked
   down, wasn't aware (or was once and forgot), and good that's
   arranged how it is

2. getters like `getpwnam` are very precious (for having a power of
   authorizing someone to something), so sanity of their usage shall
   be preserved as much as possible, and that's only possible with `_r`
   variant (side effects are bad, hopefully you'll agree) -- we can
   accidentally inflict quite a non-straightforward problem on us even
   without threads and signals, just with standard linear flow if some
   innocently looking calls get between the place of actual get and
   the use of the respective value that can be effectively overwritten
   out of our immediate sight (imposed with such new calls), e.g.:

   https://github.com/IIJ-NetBSD/netbsd-src/commit/9d235d208eb93acaa7058cf7fcebffc88181e804

3. the respective value is usually cached by the call site so even if
   there's any performance impact at all, it's nothing in the long run

Therefore I indeed think turning pre-existing `getpwnam_r` into plain
`getpwnam` would be a step back.  The direction shall be opposite for
such delicate (security sensitive + hidden side-effects) things.

Reply:
Would also note that much more libc functions are affected, but those
are usually not that critical.  E.g., would be better if `crm_perror`
used `strerror_r` where possible (it's not that portable; it may seem
OK to log something directly from the signal handler), but it both can
wait (is not security-related), and we are not actively turning into the
opposite direction, as you've suggested with `getpwnam_r` -> `getpwnam`.